runtime/src/prioritization_fee_cache.rs

```diff
@@ -0,0 +1,811 @@
+        datapoint_info!(
+            "block_prioritization_fee_counters",
+            ("slot", slot as i64, i64),
+            (
+                "successful_transaction_update_count",
+                self.successful_transaction_update_count
+                    .swap(0, Ordering::Relaxed) as i64,
+                i64
+            ),
+            (
+                "fail_get_transaction_priority_details_count",
+                self.fail_get_transaction_priority_details_count
+                    .swap(0, Ordering::Relaxed) as i64,
+                i64
+            ),
+            (
+                "fail_get_transaction_account_locks_count",
+                self.fail_get_transaction_account_locks_count
+                    .swap(0, Ordering::Relaxed) as i64,
+                i64
+            ),
+            (
+                "total_update_elapsed_us",
+                self.total_update_elapsed_us.swap(0, Ordering::Relaxed) as i64,
+                i64
+            ),
+            (
+                "total_cache_lock_elapsed_us",
+                self.total_cache_lock_elapsed_us.swap(0, Ordering::Relaxed) as i64,
+                i64
+            ),
+            (
+                "total_entry_lock_elapsed_us",
+                self.total_entry_lock_elapsed_us.swap(0, Ordering::Relaxed) as i64,
+                i64
+            ),
+            (
+                "total_evict_old_blocks_elapsed_us",
+                self.total_evict_old_blocks_elapsed_us
+                    .swap(0, Ordering::Relaxed) as i64,
+                i64
+            ),
+        );
+    }
+}
+
+/// Each block's PrioritizationFee entry is wrapped in Arc<Mutex<...>>
+/// Reader and writer should avoid to contend `PrioritizationFeeCache` but on individual block's PrioritizationFeeEntry.
+/// Each entry is assigned a unique incremental sequence_number, which is used to enforce eviction
+/// policy.
+struct PrioritizationFeeEntry {
+    entry: Arc<Mutex<PrioritizationFee>>,
+    sequence_number: u64,
+}
+
+impl PrioritizationFeeEntry {
+    pub fn new(entry: Arc<Mutex<PrioritizationFee>>, sequence_number: u64) -> Self {
+        PrioritizationFeeEntry {
+            entry,
+            sequence_number,
+        }
+    }
+
+    pub fn entry(&self) -> Arc<Mutex<PrioritizationFee>> {
+        self.entry.clone()
+    }
+
+    pub fn sequence_number(&self) -> u64 {
+        self.sequence_number
+    }
+}
+
+enum FinalizingSerivceUpdate {
+    BankFrozen {
+        slot: Slot,
+        prioritization_fee: Arc<Mutex<PrioritizationFee>>,
+    },
+    Exit,
+}
+
+/// Stores up to MAX_NUM_RECENT_BLOCKS recent block's prioritization fee,
+/// A separate internal thread `finalizing_thread` handles additional tasks when a bank is frozen,
+/// includes pruning PrioritizationFee's HashMap, collecting stats and reporting metrics.
+pub struct PrioritizationFeeCache {
+    cache: RwLock<HashMap<Slot, Arc<PrioritizationFeeEntry>>>,
+    current_sequence_number: AtomicU64,
```

Comment:
I think we should use `LruCache` like we do in the rest of the codebase like here: https://github.com/behzadnouri/solana/blob/17d9aee5202f8ae03462825e738e1db4a7fedf1e/core/src/cluster_nodes.rs#L67, instead of building our own LRU mechanism

Reply:
Impllemented in separate PR https://github.com/taozhu-chicago/solana/pull/7